kimchi/src/circuits/polynomials/foreign_field_add/circuitgates.rs

```diff
@@ -0,0 +1,233 @@
+//~ `result_carry_i` $= -1, 0, 1$ are auxiliary variables that handle carries between limbs
+//~
+//~ Apart from the range checks of the chained inputs, we need to do an additional range check for a final bound
+//~ to make sure that the result is less than the modulus, by adding `2^{3*88} - foreign_modulus` to it.
+//~ (This can be computed easily from the limbs of the modulus)
+//~ Note that `2^{264}` as limbs represents: (0, 0, 0, 1) then:
+//~
+//~ The upper-bound check can be calculated as
+//~ - `bound_lo = result_lo - foreign_modulus_lo - bound_carry_lo * 2^{88}`
+//~ - `bound_mi = result_mi - foreign_modulus_mi - bound_carry_mi * 2^{88} + bound_carry_lo`
+//~ - `bound_hi = result_hi - foreign_modulus_hi + 2^{88} + bound_carry_mi`
+//~
+//~ `bound_carry_i` $= 0$ or $1$ or $-1$ are auxiliary variables that handle carries between limbs
+//~
+//~ The range check of `bound` can be skipped until the end of the operations
+//~ and `result` is an intermediate value that is unused elsewhere (since the final `result`
+//~ must have had the right number of moduli subtracted along the way).
+//~
+//~ You could lay this out as a double-width gate for chained foreign additions and a final row, e.g.
+//~
+//~ | col | `ForeignFieldAdd`       | more `ForeignFieldAdd` | or `ForeignFieldFin` |
+//~ | --- | ----------------------- | ---------------------- | -------------------- |
+//~ |   0 | `left_input_lo`  (copy) | `result_lo` (copy)     | `resmin_lo` (copy)   |
+//~ |   1 | `left_input_mi`  (copy) | `result_mi` (copy)     | `resmin_mi` (copy)   |
+//~ |   2 | `left_input_hi`  (copy) | `result_hi` (copy)     | `resmin_hi` (copy)   |
+//~ |   3 | `right_input_lo` (copy) |  ...                   | `bound_lo`  (copy)   |
+//~ |   4 | `right_input_mi` (copy) |  ...                   | `bound_mi`  (copy)   |
+//~ |   5 | `right_input_hi` (copy) |  ...                   | `bound_hi`  (copy)   |
+//~ |   6 | `field_overflow`        |  ...                   |  -                   |
+//~ |   7 | `carry_lo`              |  ...                   | `bound_carry_lo`     |
+//~ |   8 | `carry_mi`              |  ...                   | `bound_carry_mi`     |
+//~ |   9 | `sign`                  |  ...                   |  -                   |
+//~ |  10 |                         |                        |                      |
+//~ |  11 |                         |                        |                      |
+//~ |  12 |                         |                        |                      |
+//~ |  13 |                         |                        |                      |
+//~ |  14 |                         |                        |                      |
+//~
+//~ Having a specific final row that checks the bound is useful as it checks the upper bound
+//~ without reusing the same constraints in the `ForeignFieldAdd` rows (which would require
+//~ some extra constraints to be added to the circuit).
+
+/// Implementation of the ForeignFieldAddition gate
```

Comment:
nit "Implementation of the foreign field addition gate"

I think you should comment on the high level what rows this gate works on (i.e. Curr) and that its results get output to the next row.

Reply:
ok thanks!